src/macros-by-example.md

```diff
@@ -24,149 +24,404 @@
+occurrence, it cannot be used with a separator.
+
+The repeated fragment both matches and transcribes to the specified number of
+the fragment, separated by the separator token. Metavariables are matched to
+every repetition of their corresponding fragment, so for instance, the `$(
+$i:ident ),*` example above matches `$i` to all of the identifiers in the list.
+
+During transcription, additional restrictions apply to reptitions so that the
+compiler knows how to expand them properly. First, a metavariable must appear in
+exactly the same number, kind, and nesting order of repetitions in the
+transcriber as it did in the matcher. So for the matcher `$( $i:ident ),*`, the
+transcribers `=> $i`, `=> $( $( $i)* )*`, and `=> $( $i )+` are all illegal, but
+`=> $( $i );*` is correct and replaces a comma-separated list of identifiers
+with a semicolon-separated list.
+
+Second, each repetition in the transcriber must contain at least one
+metavariable in order to decide now many times to expand it. If multiple
+metavariables appear in the same repetition, they must be bound to the same
+number of fragments. For instance, `( $( $i:ident ),* ; $( $j:ident ),* ) => (
+$( ($i,$j) ),*` must bind the same number of `$i` fragments as `$j` fragments.
+This means that invoking the macro with `(a, b, c; d, e, f`) is legal and
+expands to `((a,d), (b,e), c,f))`, but `(a, b, c; d, e)` is illegal because it
+does not have the same number. This requirement applies to every layer of nested
+repetitions.
+
+> **Edition Differences**: The `?` repetition operator did not exist before the
+> 2018 edition. Prior to the 2018 Edition, `?` was an allowed
+> separator token, rather than a repetition operator.
+
+## Scoping, Exporting, and Importing
+
+For historical reasons, the scoping of macros by example does not work entirely like
+items. Macros have two forms of scope: textual scope, and path-based scope.
+Textual scope is based on the order that things appear in source files, or even
+across multiple files, and is the default scoping. It's explained further below.
+Path-based scope works exactly the same way that item scoping does. The scoping,
+exporting, and importing of macros is controlled largely by attributes.
+
+When a macro is invoked by an unqualified identifier (not part of a multi-part
+path), it's first looked up in textual scoping. If this does not yield any
+results, then it is looked up in path-based scoping. If the macro's name is
+qualified with a path, then it is only looked up in path-based scoping.
+
+```ignore
+use lazy_static::lazy_static; // Path-based import.
+
+macro_rules! lazy_static { // Textual definition.
+    (lazy) => {};
+}
+
+lazy_static!{lazy} // Textual lookup finds our macro first.
+self::lazy_static!{} // Path-based lookup ignores our macro, finds imported one.
+```
+
+### Textual Scope
+
+Textual scope is based largely on the order that things appear in source files.
```

Comment:
IMO, It's important to draw an analogy with `let` since most of users have intuition about how `let` works, and `macro_rules` behave like "`let` at module level".